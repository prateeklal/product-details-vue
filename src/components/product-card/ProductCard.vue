<template>
  <article>
    <figure>
      <router-link :to="paramsForCarousel">
        <img
          v-lazy="hero.href"
          loading="lazy"
          :alt="hero.alt"
          :width="hero.width"
          aria-label="Main image of the product"
        />
      </router-link>
      <div class="thumbnail" @mouseleave="resetHeroImage">
        <img
          v-for="(image, index) in product.images"
          :key="index"
          :src="image.href"
          :alt="image.alt"
          @mouseenter="updateHeroImage(image)"
        />
      </div>
    </figure>

    <div class="row article-body">
      <div class="col-8">
        <router-link :to="paramsForCarousel">
          <h3 v-html="product.name"></h3>
        </router-link>
        <p tabindex="0">
          <span>Regular Price - {{ regularPrice | currency }}</span>
        </p>
      </div>
      <div class="col-4 price">
        <em tabindex="0">
          <span>Now Price</span>
          <strong>{{ nowPrice | currency }}</strong>
        </em>
        <button aria-label="Add to cart" @click="addToCart">
          <v-icon name="shopping-cart"></v-icon>
        </button>
      </div>
    </div>
  </article>
</template>

<script>
export default {
  props: ["product"],

  data() {
    return {
      hero: this.product.hero
    };
  },

  methods: {
    updateHeroImage(img) {
      this.hero = img;
    },
    resetHeroImage() {
      this.hero = this.product.hero;
    },
    addToCart() {
      const cartItems = {
        id: this.product.id,
        name: this.product.name,
        price: this.nowPrice,
        thumb: this.product.thumbnail
      };

      this.$emit("addToCart", cartItems);
    }
  },

  computed: {
    nowPrice() {
      return (
        this.product.priceRange &&
        this.product.priceRange.selling &&
        this.product.priceRange.selling.low
      );
    },
    regularPrice() {
      return this.product.priceRange.regular
        ? this.product.priceRange.regular.high
        : this.product.priceRange.selling.high;
    },
    paramsForCarousel() {
      return {
        name: "product-carousel",
        params: {
          id: this.product.id,
          images: [this.hero, ...this.product.images]
        }
      };
    }
  }
};
</script>

<style lang="scss" scoped>
@import "../../scss/_variables";

article {
  background-color: #fff;
  box-shadow: 0 3px 6px rgba(0, 0, 0, 0.1);
}

img {
  width: 100%;
}

h3 {
  font-size: em(13);
  font-weight: 500;
  line-height: em(22);
}

p {
  font-size: em(12);
  margin-top: em(15);
}

.thumbnail {
  position: absolute;
  bottom: 0;
  width: 100%;
  text-align: center;
  padding: em(5) 0;
  opacity: 0.2;
  transition: opacity 0.3s;
}

.thumbnail img {
  width: 34px;
  margin: 0 em(5);
  border: 2px solid #fff;
  border-radius: 50%;
  cursor: pointer;
}

figure {
  position: relative;
}

figure:hover .thumbnail {
  opacity: 1;
}

.article-body {
  padding: em(10) em(15) em(15);
}

em {
  font-style: normal;
  display: block;
  line-height: em(18);
}

em span {
  display: block;
  font-size: em(10);
}

em strong {
  color: #bf8b2f;
  font-size: em(17);
}

.price {
  text-align: right;
}

button {
  width: 56px;
  padding: 5px 0;
  border-radius: 5px;
  border: 1px solid $primary-color;
  margin-top: em(12);
  cursor: pointer;
  background-color: #fff;
}

button:hover {
  background-color: $primary-color;
}

button:hover .icon {
  color: #fff;
}
</style>
