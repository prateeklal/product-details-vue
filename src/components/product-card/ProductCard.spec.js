import { shallowMount } from '@vue/test-utils';
import sinon from 'sinon';
import ProductCard from './ProductCard.vue';
const localVue = global.localVue;
const router = global.router;


describe('ProductCard', () => {
  let wrapper;
  let sandbox = sinon.createSandbox();

  beforeEach(() => {
    wrapper = shallowMount(ProductCard, {
      localVue,
      router,
      propsData: {
        product: { "id": "belgian-linen-embroidery-duvet-cover-shams-slate-white-b3024", "name": "Belgian Linen Ladder Stripe Embroidery Duvet Cover &amp; Shams - Slate/White", "links": { "www": "https://www.westelm.com/products/belgian-linen-embroidery-duvet-cover-shams-slate-white-b3024/" }, "price": { "regular": 329, "selling": 44 }, "thumbnail": { "size": "m", "meta": "", "alt": "", "rel": "thumbnail", "width": 363, "href": "https://www.westelm.com/weimgs/rk/images/wcm/products/202003/0005/belgian-linen-ladder-stripe-embroidery-duvet-cover-shams-s-m.jpg", "height": 363 }, "hero": { "size": "m", "meta": "", "alt": "", "rel": "hero", "width": 363, "href": "https://www.westelm.com/weimgs/rk/images/wcm/products/202003/0005/belgian-linen-ladder-stripe-embroidery-duvet-cover-shams-s-m.jpg", "height": 363 }, "images": [{ "size": "m", "meta": "", "alt": "", "rel": "althero", "width": 363, "href": "https://www.westelm.com/weimgs/rk/images/wcm/products/202003/0005/belgian-linen-ladder-stripe-embroidery-duvet-cover-shams-6-m.jpg", "height": 363 }, { "size": "m", "meta": "", "alt": "", "rel": "althero", "width": 363, "href": "https://www.westelm.com/weimgs/rk/images/wcm/products/202003/0012/belgian-linen-ladder-stripe-embroidery-duvet-cover-shams-m.jpg", "height": 363 }], "swatches": [], "messages": [], "flags": [{ "bopisSuppress": false, "rank": 3, "id": "newcore" }, { "bopisSuppress": false, "rank": 9, "id": "fairTrade" }], "reviews": { "recommendationCount": 0, "likelihood": 0, "reviewCount": 0, "averageRating": 0, "id": "belgian-linen-embroidery-duvet-cover-shams-slate-white-b3024", "type": "GROUP_REVIEWS" } }
      }
    });
    wrapper.setData({
      hero: { "size": "m", "meta": "", "alt": "", "rel": "hero", "width": 363, "href": "https://www.westelm.com/weimgs/rk/images/wcm/products/202003/0005/belgian-linen-ladder-stripe-embroidery-duvet-cover-shams-s-m.jpg", "height": 363 }
    });
  });

  afterEach(() => {
    sandbox.restore();
  });

  it('should be a vue instance', () => {
    expect(wrapper.isVueInstance()).toBeTruthy()
  });

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should emit addtocart with correct cartitems object', () => {
    wrapper.find('button').trigger('click');
    const emittedObj = wrapper.emitted().addToCart[0][0];
    const expectedProps = ['id', 'name', 'price', 'thumb'];
    const receivedProps = Object.keys(emittedObj);
    expect(receivedProps).toEqual(expectedProps);
  });

  it('should update hero image on mouseover', () => {
    sandbox.spy(wrapper.vm, 'updateHeroImage');
    wrapper.find('.thumbnail img').trigger('mouseenter');
    expect(wrapper.vm.updateHeroImage.calledOnce).toBeTruthy();
    expect(wrapper.vm.$data.hero.href).toEqual(wrapper.vm.updateHeroImage.getCall(0).args[0].href)
  });

  it('should update hero image on mouseleave', () => {
    wrapper.vm.resetHeroImage();
    expect(wrapper.vm.$data.hero.href).toEqual(wrapper.vm.product.hero.href);
  });
});