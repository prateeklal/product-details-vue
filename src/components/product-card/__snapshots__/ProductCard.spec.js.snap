// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard renders correctly 1`] = `
<article>
  <figure>
    <router-link to="[object Object]"><img loading="lazy" alt="" width="363" aria-label="Main image of the product"></router-link>
    <div class="thumbnail"><img src="https://www.westelm.com/weimgs/rk/images/wcm/products/202003/0005/belgian-linen-ladder-stripe-embroidery-duvet-cover-shams-6-m.jpg" alt=""><img src="https://www.westelm.com/weimgs/rk/images/wcm/products/202003/0012/belgian-linen-ladder-stripe-embroidery-duvet-cover-shams-m.jpg" alt=""></div>
  </figure>
  <div class="row article-body">
    <div class="col-8">
      <router-link to="[object Object]">
        <h3>Belgian Linen Ladder Stripe Embroidery Duvet Cover &amp; Shams - Slate/White</h3>
      </router-link>
      <p tabindex="0"><span>Regular Price - 329</span></p>
    </div>
    <div class="col-4 price"><em tabindex="0"><span>Now Price</span> <strong>35.2</strong></em> <button aria-label="Add to cart">
        <v-icon name="shopping-cart"></v-icon>
      </button></div>
  </div>
</article>
`;
